all-sdk/autotuning.md

```diff
@@ -0,0 +1,452 @@
+    W->>W: Cache new workflow
+  end
+  
+  W->>S: `mark_slot_used`
+  W->>W: Process workflow task
+  W->>S: `release_slot`
+```
+
+## Default implementations
+
+We should provide a few default implementations:
+
+* Fixed-size SlotSupplier: This is the current implementation. There's a fixed number of slots, that's it.
+* Resource based SlotSupplier: This implementation can look at how close the worker is to some resource limit
+  (% of CPU and/or memory used), and hand out slots if it's still below that limit. This only makes sense for
+  activity slot suppliers, since workflows use little CPU, and the memory they use is the concern of cache
+  management, not the slot supplier.
+* Memory based WorkflowCacheSizer: Like above, but for the sizing of the workflow cache.
+
+## Metrics
+
+Users can add their own metrics behind their implementations now, which is great - but we can also provide some out of
+the box. We can always provide `slots_in_use` by type. Right now we emit the inverse of this,
+`worker_task_slots_available` - we can keep emitting that with a bundled implementation that replicates the old
+fixed-number based implementation. 
+
+We can also of course keep emitting the current number of cached workflows.
+
+
+## Where it hooks up and the language barrier
+
+Staying in Rust for a moment, the slot supplier implementations can be hooked up on a per-worker basis through
+`WorkerConfig`, as expected. We would add:
+
+```rust
+pub struct WorkerConfig {
+  // ... existing fields
+  #[builder(default)]
+  pub workflow_task_slot_and_cache: Option<Arc<dyn WorkflowSlotAndCacheManager>>,
+  #[builder(default)]
+  pub activity_task_slot_supplier: Option<Arc<dyn ActivityTaskSlotSupplier>>,
+  #[builder(default)]
+  pub local_activity_task_slot_supplier: Option<Arc<dyn LocalActivityTaskSlotSupplier>>,
+}
+
+impl WorkerConfig {
+  /// Set the slot supplier for workflow tasks, using the default fixed-size workflow cache
+  pub fn with_workflow_slot_supplier(&mut self, supplier: impl WorkflowTaskSlotSupplier) -> &mut Self { /* ... */ }
+}
+```
+
+This is analogous to what we'd expect in the worker config for other languages.
+
+The suppliers are kept inside of `Arc`s so that they can be shared across multiple workers, and used with the pauseable
+supplier (which the user presumably wants to retain a reference to so they can call pause).
+
+Of course, most users aren't going to be implementing this in Rust, they'll be doing it in their language, and then
+we've got to translate that into calls across the language barrier. As such we need a slot provider implementation
+that exists in each language's bridge, and calls callbacks into the user's implementation.
+
+Something like:
+
+```rust
+pub struct RustToLangBridgeSlotSupplier {
+  // ... store callbacks etc
+}
+
+impl<T: SlotKind> SlotSupplier<SlotKind=T> for RustToLangBridgeSlotSupplier {
+  // ... call appropriate callbacks, convert data types, etc
+}
+```
+
+## A neat example
```

Comment:
This is neat, many of our users keep asking how they can limit activities, so being able to acquire a zookeeper/etcd lock for a specific activity types is helpful. But we're gonna have to provide helpers if we want to advertise support for this, it's too complicated for most users.